### PR TITLE
Fire Initialized/BeforeDestroyed/Destroyed events for request context

### PR DESCRIPTION
**Pull request by manovotn**
If I am not mistaken then you can handle context activation manually, via interceptor and via controller bean. All of these should be incorporated and tested in this PR. Feedback and nagging are welcome ;-)

### CODE REVIEW

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/RequestContext.java [hunk 1]
```diff
@@ -104,6 +110,11 @@ public void activate(Collection<ContextInstanceHandle<?>> initialState) {
             }
         }
         currentContext.set(state);
+        // Fire an event with qualifier @Initialized(RequestScoped.class)
+        Set<Annotation> qualifiers = new HashSet<>(4);
+        qualifiers.add(Initialized.Literal.REQUEST);
+        qualifiers.add(Any.Literal.INSTANCE);
+        EventImpl.createNotifier(Object.class, Object.class, qualifiers, ArcContainerImpl.instance()).notify(toString());
```
Comment (mkouba) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/RequestContext.java]: I think we should cache non-empty notifiers and skip notifications for empty notifiers.

**Review by mkouba — CHANGES_REQUESTED** [review]
Otherwise looks good.

**Comment by manovotn** [comment]
@mkouba updated, thanks for the tip with `LazyValue`!

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/RequestContext.java [hunk 2]
```diff
@@ -118,7 +144,16 @@ public void destroy() {
             synchronized (ctx) {
                 for (InstanceHandle<?> instance : ctx.values()) {
                     try {
+                        ArcContainerImpl arcContainer = ArcContainerImpl.instance();
```
Comment (mkouba) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/RequestContext.java]: `arcContainer` is not used...

Comment (manovotn) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/RequestContext.java]: Removed now

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/RequestContext.java [hunk 3]
```diff
@@ -104,6 +126,10 @@ public void activate(Collection<ContextInstanceHandle<?>> initialState) {
             }
         }
         currentContext.set(state);
+        // Fire an event with qualifier @Initialized(RequestScoped.class) if there are any observers for it
+        if (!initializedNotifier.get().isEmpty()) {
```
Comment (mkouba) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/RequestContext.java]: `LazyValue.get()` is always a `volatile` read and we should try to make it as fast as possible. I'd probably create a util method like this:
```java
static void fireIfNotEmpty(LazyValue<Notifier<Object>> value) {
   Notifier notifier = value.get();
   if (!notifier.isEmpty()) {
      notifier.notify(toString());
   }
}
```
WDYT?

**Review by mkouba — CHANGES_REQUESTED** [review]

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/RequestContext.java (hunk 3 above)
Comment (manovotn) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/RequestContext.java]: That's fair point, I've altered the code.
Though I am not sure why do you want the util method to be `static`?

Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/RequestContext.java]: Ah sorry, of course it doesn't have to be `static`...

Comment (manovotn) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/RequestContext.java]: Thought so ;)
It's private now and hence doesn't need the secondary parameter.

**Review by mkouba — APPROVED** [review]
Looks good now. Let's wait for CI...

**Pull request merged**